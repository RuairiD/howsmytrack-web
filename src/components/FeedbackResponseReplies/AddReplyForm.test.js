import React from "react";
import { mount, shallow } from "enzyme";
import toJson from "enzyme-to-json";
import axios from "axios";
import waitForExpect from "wait-for-expect";
import { act } from "react-dom/test-utils";
import AddReplyForm from "./AddReplyForm";

jest.mock("axios");

describe("AddReplyForm", () => {
    beforeEach(() => {
        axios.post.mockRestore();
    });

    it("renders an empty text field if replies are allowed for this thread", () => {
        const wrapper = shallow(
            <AddReplyForm
                feedbackResponseId={1901}
                refetchReplies={jest.fn()}
                allowFurtherReplies
            />,
        );
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it("renders an an informational message if replies are not allowed for this thread", () => {
        const wrapper = shallow(
            <AddReplyForm
                feedbackResponseId={1901}
                refetchReplies={jest.fn()}
            />,
        );
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it("renders nothing if replies are not allowed, but replies are still loading", () => {
        const wrapper = shallow(
            <AddReplyForm
                feedbackResponseId={1901}
                refetchReplies={jest.fn()}
                isLoadingReplies
            />,
        );
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it("posts a new reply if one is submitted, refetches replies upstream, and resets the form", async () => {
        axios.post.mockResolvedValue({
            data: {
                data: {
                    addFeedbackResponseReply: {
                        reply: {
                            allowReplies: true,
                        },
                        error: null,
                    },
                },
            },
        });

        const refetchReplies = jest.fn();
        const wrapper = mount(
            <AddReplyForm
                feedbackResponseId={1901}
                refetchReplies={refetchReplies}
                allowFurtherReplies
            />,
        );

        await act(async () => wrapper.find("TextArea").get(0).props.onChange({
            target: {
                value: "This is a new reply.",
            },
        }));
        await act(async () => wrapper.find("Checkbox").get(0).props.onChange({
            target: {
                value: true,
            },
        }));

        await act(async () => wrapper.find("Button").get(0).props.onClick());

        await waitForExpect(() => {
            wrapper.update();
            expect(refetchReplies).toHaveBeenCalled();
            expect(wrapper.find("TextArea").get(0).props.value).toBe("");
        });
    });

    it("prevents further replies if the user disabled replies when sending a new reply", async () => {
        axios.post.mockResolvedValue({
            data: {
                data: {
                    addFeedbackResponseReply: {
                        reply: {
                            allowReplies: false,
                        },
                        error: null,
                    },
                },
            },
        });

        const wrapper = mount(
            <AddReplyForm
                feedbackResponseId={1901}
                refetchReplies={jest.fn()}
                allowFurtherReplies
            />,
        );

        await act(async () => wrapper.find("TextArea").get(0).props.onChange({
            target: {
                value: "This is the final reply. Good day.",
            },
        }));
        await act(async () => wrapper.find("Checkbox").get(0).props.onChange({
            target: {
                value: false,
            },
        }));

        await act(async () => wrapper.find("Button").get(0).props.onClick());

        await waitForExpect(() => {
            wrapper.update();
            expect(wrapper.find("TextArea").length).toBe(0);
            expect(wrapper.text().includes("Additional replies have been disabled for this conversation.")).toBeTruthy();
        });
    });

    it("shows an error if adding a new reply failed", async () => {
        axios.post.mockResolvedValue({
            data: {
                data: {
                    addFeedbackResponseReply: {
                        reply: null,
                        error: "error",
                    },
                },
            },
        });

        const wrapper = mount(
            <AddReplyForm
                feedbackResponseId={1901}
                refetchReplies={jest.fn()}
                allowFurtherReplies
            />,
        );

        await act(async () => wrapper.find("TextArea").get(0).props.onChange({
            target: {
                value: "This is the final reply. Good day.",
            },
        }));

        await act(async () => wrapper.find("Button").get(0).props.onClick());

        await waitForExpect(() => {
            wrapper.update();
            expect(wrapper.find("Alert").length).toBe(1);
        });
    });
});
