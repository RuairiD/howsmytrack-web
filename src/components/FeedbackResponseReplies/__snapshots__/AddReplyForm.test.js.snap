// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddReplyForm renders an an informational message if replies are not allowed for this thread 1`] = `
<Div
  className=""
  style={
    Object {
      "padding": "0.5em 0",
    }
  }
>
  <Text
    strong={true}
  >
    Additional replies have been disabled for this conversation.
  </Text>
</Div>
`;

exports[`AddReplyForm renders an empty text field if replies are allowed for this thread 1`] = `
<Div
  className=""
  style={
    Object {
      "padding": "0.5em 0",
    }
  }
>
  <Div
    className=""
  >
    <Row
      gutter={
        Array [
          16,
          16,
        ]
      }
    >
      <Col />
    </Row>
    <Row
      gutter={
        Array [
          16,
          16,
        ]
      }
    >
      <Col>
        <TextArea
          autoFocus={true}
          onChange={[Function]}
          placeholder="Write your reply..."
          rows={2}
          value=""
        />
      </Col>
    </Row>
    <Row
      gutter={
        Array [
          16,
          16,
        ]
      }
    >
      <Col
        style={
          Object {
            "display": "flex",
          }
        }
      >
        <Checkbox
          checked={true}
          indeterminate={false}
          onChange={[Function]}
          style={
            Object {
              "marginRight": "auto",
            }
          }
        >
          Allow additional replies.
        </Checkbox>
        <Button
          block={false}
          disabled={true}
          ghost={false}
          htmlType="button"
          loading={false}
          onClick={[Function]}
          style={
            Object {
              "marginLeft": "auto",
            }
          }
          type="primary"
        >
          Send Reply
        </Button>
      </Col>
    </Row>
  </Div>
</Div>
`;

exports[`AddReplyForm renders nothing if replies are not allowed, but replies are still loading 1`] = `
<Div
  className=""
  style={
    Object {
      "padding": "0.5em 0",
    }
  }
/>
`;
